torch/testing/_internal/common_methods_invocations.py

```diff
@@ -4,12 +4,122 @@
 from operator import mul, itemgetter
 import collections
 from torch.autograd import Variable
-from torch.testing import make_non_contiguous
-from torch.testing._internal.common_device_type import skipCUDAIfNoMagma, skipCPUIfNoLapack, expectedFailureCUDA
-from torch.testing._internal.common_utils import (prod_single_zero, random_square_matrix_of_rank,
-                                                  random_symmetric_matrix, random_symmetric_psd_matrix,
-                                                  random_symmetric_pd_matrix, make_nonzero_det,
-                                                  random_fullrank_matrix_distinct_singular_value, set_rng_seed)
+from torch.testing import \
+    (make_non_contiguous,
+     get_common_float_types, get_all_float_types,
+     get_common_float_and_complex_types_plus_half,
+     get_common_float_and_complex_types_plus_bfloat16,
+     get_all_float_and_complex_types)
+from torch.testing._internal.common_device_type import \
+    (skipCUDAIfNoMagma, skipCPUIfNoLapack, expectedFailureCUDA)
+from torch.testing._internal.common_utils import \
+    (prod_single_zero, random_square_matrix_of_rank,
+     random_symmetric_matrix, random_symmetric_psd_matrix,
+     random_symmetric_pd_matrix, make_nonzero_det,
+     random_fullrank_matrix_distinct_singular_value, set_rng_seed)
+
+
+# Classes and methods for the operator database
+# The OpMeta class contains metadata and helpers for acquiring properties of
+# operations, like what dtypes they support.
+class OpMeta(object):
+    def __init__(self,
+                 name,  # the string name of the function
+                 *,
+                 dtypes=None,  # dtypes this function is expected to work with
+                 dtypesIfCPU=None,  # dtypes this function is expected to work with on CPU
+                 dtypesIfCUDA=None,  # dtypes this function is expected to work with on CUDA
+                 dtypesIfROCM=None,  # dtypes this function is expected to work with on ROCM
+                 has_out_kwarg=None,  # whether this function has an out kwarg
+                 decorators=None,  # decorators to apply to generated tests
+                 tests_to_skip=None):  # regular expressions describing tests to skip
```

Comment:
What use cases do you anticipate a regex is better than just a plain old skip list? It's harder to check if a regex is bitrotting, whereas a skip list we can complain if it refers to a test that doesn't exist.

Reply:
That is true, but we generate so many variants of a test I found it very convenient to move to a regex system when implementing additional operators.

Reply:
I remove skipping from this PR since it's unused.